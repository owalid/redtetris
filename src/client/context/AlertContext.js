import createDataContext from "./createDataContext";
export const ALERT_POP = 'ALERT_POP'

const initialState = {
  message: '',
  type: null,
};

const alertReducer = (state, action) => {
  switch(action.type){
    case ALERT_POP:
      return { message: action.message, type: action.typeAlert }
    default: 
      return state
  }
};

const sendAlert = (dispatch) => (message, typeAlert) =>
  dispatch({ type: ALERT_POP, typeAlert, message })

export const { Provider, Context } = createDataContext(
  //reducer :
  alertReducer,

  //action functions :
  {
    sendAlert
  },

  //initialState :
  {
   ...initialState
  }
);