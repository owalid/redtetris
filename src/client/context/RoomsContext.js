import createDataContext from "./createDataContext";

const initialState = {
  rooms: null
};

const RoomsReducer = (state, payload) => {
  switch (payload.type) {
    case "updateRooms":
      return { ...state, rooms: payload.rooms };
    default:
      return state;
  }
};

const updateRooms = dispatch => rooms => {
  if (Object.keys(rooms).includes("_data")) {
    dispatch({ type: "updateRooms", rooms: rooms._data });
  } else {
    dispatch({ type: "updateRooms", rooms: rooms });
  }
}

export const { Provider, Context } = createDataContext(
  //reducer :
  RoomsReducer,

  //action functions :
  {
    updateRooms
  },

  //initialState :
  {
   ...initialState
  }
);