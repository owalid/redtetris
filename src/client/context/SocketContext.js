import React, { createContext } from "react"
import socketIOClient from "socket.io-client"
require('dotenv').config() 

export const SocketContext = createContext();

export const SocketContextProvider = ({ children }) => {
  const socketClient = socketIOClient(process.env.API_URL);  
  const sendSocket = (type, data = false) => {
    socketClient.emit(type, data);
  }
  
  return (
    <SocketContext.Provider
      value={{ socketClient, sendSocket }}
    >
      { children }
    </SocketContext.Provider>
  )
}