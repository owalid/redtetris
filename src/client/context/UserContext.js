import createDataContext from "./createDataContext";

const initialState = {
  uuidRoom: null,
  player: null
};

const UserReducer = (state, payload) => {
  switch (payload.type) {
    case "updateUuidRoom": {
      return {  ...state, uuidRoom: payload.uuidRoom };
    }
    case "updatePlayer": 
      return {  ...state, player: payload.player };
    default:
      return state;
  }
};

const updateUuidRoom = dispatch => uuidRoom => 
  dispatch({ type: "updateUuidRoom", uuidRoom });

const updatePlayer = dispatch => player => 
  dispatch({ type: "updatePlayer", player });

export const { Provider, Context } = createDataContext(
  //reducer :
  UserReducer,

  //action functions :
  {
    updateUuidRoom,
    updatePlayer
  },

  //initialState :
  {
   ...initialState
  }
);