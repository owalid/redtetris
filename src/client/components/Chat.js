import React, { useState, useContext, useEffect } from 'react';
import { Button, Input } from '@material-ui/core';

import { Context as RoomsContext } from "../context/RoomsContext";
import { SocketContext } from "../context/SocketContext";
import { Context as UserContext } from "../context/UserContext";

const Chat = ({ uuidRoom }) => {
  const { sendSocket } = useContext(SocketContext);
  const { state: { player } } = useContext(UserContext);
  const { state: { rooms } } = useContext(RoomsContext);
  const [message, setMessage] = useState('');
  const [game, setGame] = useState(true);
  const keysCode = ['ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight', ' ', 'Escape'];

  useEffect(() => {
    const getKeyUp = (event) => {
      const { key } = event;
      if (keysCode.includes(key)) {
        if (key === 'Escape') {
          sendSocket('server/pause-resume', { channel: uuidRoom })
        } else {
          sendSocket('server/key-up', { key, channel: uuidRoom, uuidUser: player.uuid  })
        }
      }
    }
    if (game) {
      document.addEventListener('keyup', getKeyUp);
    } else {
      document.removeEventListener('keyup', getKeyUp);
    }
    return () => {
      document.removeEventListener('keyup', getKeyUp);
    };
  }, [game])


  const sendMessage = (e) => {
    e.preventDefault();
    if (message !== "") {
      sendSocket('server/new-message', {
          content: message,
          uuidRoom,
          login: player.name,
          id_user: player.uuid
      })
      setMessage("")
      const element = document.getElementById("chat");
      element.scrollTop = 0;
    }
  }

  useEffect(() => {
    const inputMessage = document.getElementById('input-message')
    inputMessage.addEventListener('focusout', (event) => {
      setGame(true);
    });
    inputMessage.addEventListener('focus', (event) => {
      setGame(false);
    });
  }, [])

return (
  <div>
    <div id="chat" className="container--chat">
      <div className="chat d-flex fdir--column">
        {
          rooms && rooms[uuidRoom].messages.map(msg => {
            if (!msg.login) {
                return (
                  <div
                    className="test--message-base"
                    key={msg.uuid}
                  >
                    <p>
                      <span className={`bold medium-text ${(msg.uuidUser === -1) ? "text--green" : "text--red"}`}>
                        {msg.time} -- { msg.content }
                      </span><br />
                    </p>
                  </div>
                )
              } else {
                  return (
                    <div
                      key={msg.uuid}
                      className={`${msg.uuidUser === player.uuid
                        ? 'aself--fend bubble--me'
                        : 'aself--fstart bubble--other'}
                        aself--fstart
                        pb-2
                        bubble
                        test--message-user
                      `}
                    >
                    <p>
                      <span className="bold medium-text">{ msg.login }:</span> { msg.content } <br />
                    </p>
                    <p className="thin small-text pt-0">{msg.time}</p>
                  </div>
                )
              }
            })
          }
        </div>
      </div>
      <div className="d-flex fdir--column">

      <div className="aself--center">
        <form onSubmit={sendMessage}>
            <Input
              className="aself--center mb-2 input-size"
              value={message}
              onChange={e => setMessage(e.target.value)}
              variant="outlined"
              required
              name="message"
              label="message"
              type="message"
              id="input-message"
              inputProps={{
                'data-testid': 'messageChat'
              }}
              />
            <Button
              className="aself--center mt-2 test--btn-submit-message"
              color="primary"
              data-testid='btnLogin'
              disabled={message.length === 0}
              onClick={sendMessage}
              >
              ➡️
            </Button>
          </form>
        </div>
      </div>
    </div>
  )
}

export default Chat
