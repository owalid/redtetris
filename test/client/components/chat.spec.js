import React, { useContext, useEffect } from "react";
import '@testing-library/jest-dom/extend-expect';
import Enzyme, { shallow } from "enzyme";
import Chat from '../../../src/client/components/Chat'
import { render, fireEvent } from '@testing-library/react'
import Adapter from "enzyme-adapter-react-16";
import io from 'socket.io-client';
import { rooms_instance, visitor_player, room1, room3 } from '../helpers/data'
import { TestAppSocketProvider } from "../helpers/SocketContext";
import { TestAppUserProvider } from "../helpers/UserContext";
import { TestAppRoomsProvider } from "../helpers/RoomsContext";
import { Context as UserContext } from "../../../src/client/context/UserContext";
import { Context as RoomsContext } from "../../../src/client/context/RoomsContext";

jest.mock('socket.io-client', () => {
  const mSocket = {
    emit: jest.fn(),
  };
  return jest.fn(() => mSocket);
});
Enzyme.configure({ adapter: new Adapter() });

describe('Chat component', () => {

  const Wrapper = () => (
    <Chat />
  )

  it('Is exists', () => {
    const wrapper = shallow(<Wrapper />);
    expect(wrapper).toMatchSnapshot();
  });

  it('Can mount', () => {

    const CurrentPlayerSetter = ({ children }) => {
      const { updatePlayer } = useContext(UserContext);
      useEffect(() => {
        updatePlayer(visitor_player);
      }, [])
      return <>{children}</>;
    };

    const CurrentRoomsSetter = ({ children }) => {
      const { updateRooms } = useContext(RoomsContext);
      useEffect(() => {
        updateRooms(rooms_instance);
      }, []);
      return <>{children}</>;
    };

    const Wr = () => (
      <TestAppSocketProvider>
        <TestAppUserProvider>
          <CurrentPlayerSetter>
            <TestAppRoomsProvider>
              <CurrentRoomsSetter>
                <Chat
                  uuidRoom={room1.channel}
                  />
              </CurrentRoomsSetter>
            </TestAppRoomsProvider>
          </CurrentPlayerSetter>
        </TestAppUserProvider>
      </TestAppSocketProvider>
    )
    const w = render(<Wr />);
    expect(w).not.toBeNull()
  })

  it('Test button message', () => {
    const CurrentPlayerSetter = ({ children }) => {
      const { updatePlayer } = useContext(UserContext);
      useEffect(() => {
        updatePlayer(visitor_player);
      }, [])
      return <>{children}</>;
    };

    const CurrentRoomsSetter = ({ children }) => {
      const { updateRooms } = useContext(RoomsContext);
      useEffect(() => {
        updateRooms(rooms_instance);
      }, []);
      return <>{children}</>;
    };

    const Wr = () => (
      <TestAppSocketProvider>
        <TestAppUserProvider>
          <CurrentPlayerSetter>
            <TestAppRoomsProvider>
              <CurrentRoomsSetter>
                <Chat
                  uuidRoom={room1.channel}
                  />
              </CurrentRoomsSetter>
            </TestAppRoomsProvider>
          </CurrentPlayerSetter>
        </TestAppUserProvider>
      </TestAppSocketProvider>
    )
    const { container, getByTestId } = render(<Wr />);
    const input_messsage = getByTestId('messageChat');
    const button_submit = container.querySelector('.test--btn-submit-message')

    expect(button_submit).toBeDisabled();
    fireEvent.change(input_messsage, { target: { value: 'abcd' } });
    expect(button_submit).not.toBeDisabled();
  })

  it('Send message', () => {
    const CurrentPlayerSetter = ({ children }) => {
      const { updatePlayer } = useContext(UserContext);
      useEffect(() => {
        updatePlayer(visitor_player);
      }, [])
      return <>{children}</>;
    };

    const CurrentRoomsSetter = ({ children }) => {
      const { updateRooms } = useContext(RoomsContext);
      useEffect(() => {
        updateRooms(rooms_instance);
      }, []);
      return <>{children}</>;
    };

    const Wr = () => (
      <TestAppSocketProvider>
        <TestAppUserProvider>
          <CurrentPlayerSetter>
            <TestAppRoomsProvider>
              <CurrentRoomsSetter>
                <Chat
                  uuidRoom={room1.channel}
                  />
              </CurrentRoomsSetter>
            </TestAppRoomsProvider>
          </CurrentPlayerSetter>
        </TestAppUserProvider>
      </TestAppSocketProvider>
    )
    const { container, getByTestId } = render(<Wr />);
    const input_messsage = getByTestId('messageChat');
    const button_submit = container.querySelector('.test--btn-submit-message')
    
    fireEvent.change(input_messsage, { target: { value: 'abcd' } });
    fireEvent.click(button_submit)
  })

  it('Focus unFocus input', () => {
    const CurrentPlayerSetter = ({ children }) => {
      const { updatePlayer } = useContext(UserContext);
      useEffect(() => {
        updatePlayer(visitor_player);
      }, [])
      return <>{children}</>;
    };

    const CurrentRoomsSetter = ({ children }) => {
      const { updateRooms } = useContext(RoomsContext);
      useEffect(() => {
        updateRooms(rooms_instance);
      }, []);
      return <>{children}</>;
    };

    const Wr = () => (
      <TestAppSocketProvider>
        <TestAppUserProvider>
          <CurrentPlayerSetter>
            <TestAppRoomsProvider>
              <CurrentRoomsSetter>
                <Chat
                  uuidRoom={room1.channel}
                  />
              </CurrentRoomsSetter>
            </TestAppRoomsProvider>
          </CurrentPlayerSetter>
        </TestAppUserProvider>
      </TestAppSocketProvider>
    )
    const { container, getByTestId } = render(<Wr />);
    const input_messsage = getByTestId('messageChat');
    
    fireEvent.focus(input_messsage)
    fireEvent.focusOut(input_messsage)
  })

  it('Have base message', () => {
    const CurrentPlayerSetter = ({ children }) => {
      const { updatePlayer } = useContext(UserContext);
      useEffect(() => {
        updatePlayer(visitor_player);
      }, [])
      return <>{children}</>;
    };

    const CurrentRoomsSetter = ({ children }) => {
      const { updateRooms } = useContext(RoomsContext);
      useEffect(() => {
        updateRooms(rooms_instance);
      }, []);
      return <>{children}</>;
    };

    const Wr = () => (
      <TestAppSocketProvider>
        <TestAppUserProvider>
          <CurrentPlayerSetter>
            <TestAppRoomsProvider>
              <CurrentRoomsSetter>
                <Chat
                  uuidRoom={room1.channel}
                  />
              </CurrentRoomsSetter>
            </TestAppRoomsProvider>
          </CurrentPlayerSetter>
        </TestAppUserProvider>
      </TestAppSocketProvider>
    )
    const { container } = render(<Wr />);
    const message_base = container.querySelector('.test--message-base')
    expect(message_base).not.toBeNull();

  })

  it('Have message user', () => {
    const CurrentPlayerSetter = ({ children }) => {
      const { updatePlayer } = useContext(UserContext);
      useEffect(() => {
        updatePlayer(visitor_player);
      }, [])
      return <>{children}</>;
    };

    const CurrentRoomsSetter = ({ children }) => {
      const { updateRooms } = useContext(RoomsContext);
      useEffect(() => {
        updateRooms(rooms_instance);
      }, []);
      return <>{children}</>;
    };

    const Wr = () => (
      <TestAppSocketProvider>
        <TestAppUserProvider>
          <CurrentPlayerSetter>
            <TestAppRoomsProvider>
              <CurrentRoomsSetter>
                <Chat
                  uuidRoom={room3.channel}
                  />
              </CurrentRoomsSetter>
            </TestAppRoomsProvider>
          </CurrentPlayerSetter>
        </TestAppUserProvider>
      </TestAppSocketProvider>
    )
    const { container } = render(<Wr />);
    const message_base = container.querySelector('.test--message-user')
    expect(message_base).not.toBeNull();

  })

  it('Send message', () => {
    const CurrentPlayerSetter = ({ children }) => {
      const { updatePlayer } = useContext(UserContext);
      useEffect(() => {
        updatePlayer(visitor_player);
      }, [])
      return <>{children}</>;
    };

    const CurrentRoomsSetter = ({ children }) => {
      const { updateRooms } = useContext(RoomsContext);
      useEffect(() => {
        updateRooms(rooms_instance);
      }, []);
      return <>{children}</>;
    };

    const Wr = () => (
      <TestAppSocketProvider>
        <TestAppUserProvider>
          <CurrentPlayerSetter>
            <TestAppRoomsProvider>
              <CurrentRoomsSetter>
                <Chat
                  uuidRoom={room1.channel}
                  />
              </CurrentRoomsSetter>
            </TestAppRoomsProvider>
          </CurrentPlayerSetter>
        </TestAppUserProvider>
      </TestAppSocketProvider>
    )

    const { container, getByTestId } = render(<Wr />);
    const input_messsage = getByTestId('messageChat');
    const button_submit = container.querySelector('.test--btn-submit-message')
    fireEvent.change(input_messsage, { target: { value: 'abcd' } });
    fireEvent.click(button_submit)
    const message_user = container.querySelector('.test--message-user')
    expect(message_user).toBeNull();
  })

  it('Scape without focus', () => {
    const ENDPOINT = 'localhost:3004';
    const mockSocket = io(ENDPOINT);

    const CurrentPlayerSetter = ({ children }) => {
      const { updatePlayer } = useContext(UserContext);
      useEffect(() => {
        updatePlayer(visitor_player);
      }, [])
      return <>{children}</>;
    };

    const CurrentRoomsSetter = ({ children }) => {
      const { updateRooms } = useContext(RoomsContext);
      useEffect(() => {
        updateRooms(rooms_instance);
      }, []);
      return <>{children}</>;
    };

    const Wr = () => (
      <TestAppSocketProvider>
        <TestAppUserProvider>
          <CurrentPlayerSetter>
            <TestAppRoomsProvider>
              <CurrentRoomsSetter>
                <Chat
                  uuidRoom={room1.channel}
                  />
              </CurrentRoomsSetter>
            </TestAppRoomsProvider>
          </CurrentPlayerSetter>
        </TestAppUserProvider>
      </TestAppSocketProvider>
    )

    const { container } = render(<Wr />);
    fireEvent.keyDown(container, {
      key: "Escape",
      code: "Escape",
      keyCode: 27,
      charCode: 27
    });
    expect(mockSocket.emit).toHaveBeenCalledTimes(2)
  })

  it('Arrow without focus', () => {
    const ENDPOINT = 'localhost:3004';
    const mockSocket = io(ENDPOINT);

    const CurrentPlayerSetter = ({ children }) => {
      const { updatePlayer } = useContext(UserContext);
      useEffect(() => {
        updatePlayer(visitor_player);
      }, [])
      return <>{children}</>;
    };

    const CurrentRoomsSetter = ({ children }) => {
      const { updateRooms } = useContext(RoomsContext);
      useEffect(() => {
        updateRooms(rooms_instance);
      }, []);
      return <>{children}</>;
    };

    const Wr = () => (
      <TestAppSocketProvider>
        <TestAppUserProvider>
          <CurrentPlayerSetter>
            <TestAppRoomsProvider>
              <CurrentRoomsSetter>
                <Chat
                  uuidRoom={room1.channel}
                  />
              </CurrentRoomsSetter>
            </TestAppRoomsProvider>
          </CurrentPlayerSetter>
        </TestAppUserProvider>
      </TestAppSocketProvider>
    )

    const { container } = render(<Wr />);
    fireEvent.keyDown(container, {
      key: "ArrowRight",
      code: "ArrowRight",
      keyCode: 39,
      charCode: 39
    });
    expect(mockSocket.emit).toHaveBeenCalledTimes(2)
  })

});