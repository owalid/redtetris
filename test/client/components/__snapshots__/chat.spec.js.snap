// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat component Is exists 1`] = `<Chat />`;
